nautobot/cloud/models.py

```diff
@@ -0,0 +1,247 @@
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "webhooks",
+)
+class CloudResourceType(PrimaryModel):
+    name = models.CharField(max_length=CHARFIELD_MAX_LENGTH, help_text="Type of cloud objects", unique=True)
+    description = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True)
+    provider = models.ForeignKey(
+        to="dcim.Manufacturer",
+        on_delete=models.PROTECT,
+        related_name="cloud_resource_types",
+        help_text="The Manufacturer instance which represents the Cloud Provider",
+    )
+    config_schema = models.JSONField(null=True, blank=True)
+    content_types = models.ManyToManyField(
+        to=ContentType,
+        help_text="The content type(s) to which this model applies.",
+        related_name="cloud_resource_types",
+        limit_choices_to=FeatureQuery("cloud_resource_types"),
+    )
+    clone_fields = [
+        "provider",
+        "content_types",
+    ]
+
+    objects = BaseManager.from_queryset(ContentTypeRelatedQuerySet)()
+
+    class Meta:
+        ordering = ["name"]
+
+    def __str__(self):
+        return self.name
+
+    @property
+    def display(self):
+        return f"{self.provider}: {self.name}"
+
+
+class CloudResourceTypeMixin(models.Model):
+    """Mixin that designates a model as compatible with CloudResourceType content_types selections."""
+
+    cloud_resource_type = ForeignKeyLimitedByContentTypes(to=CloudResourceType, on_delete=models.PROTECT)
+    extra_config = models.JSONField(null=True, blank=True)
+
+    is_cloud_resource_type_model = True
+
+    class Meta:
+        abstract = True
+
+    def clean(self):
+        super().clean()
+
+        # Copied from nautobot.extras.models.models.ConfigContextSchemaValidationMixin
+        schema = self.cloud_resource_type.config_schema
+        if schema:
+            try:
+                Draft7Validator(schema, format_checker=Draft7Validator.FORMAT_CHECKER).validate(self.extra_config)
+            except JSONSchemaValidationError as e:
+                raise ValidationError(
+                    {
+                        "extra_config": [
+                            f"Validation according to CloudResourceType {self.cloud_resource_type} config_schema failed.",
+                            e.message,
+                        ]
+                    }
+                )
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "webhooks",
+)
+class CloudNetwork(CloudResourceTypeMixin, PrimaryModel):
+    name = models.CharField(max_length=CHARFIELD_MAX_LENGTH, unique=True)
+    description = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True)
+    cloud_account = models.ForeignKey(to=CloudAccount, on_delete=models.PROTECT, related_name="cloud_networks")
+    parent = models.ForeignKey(
+        to="cloud.CloudNetwork", on_delete=models.SET_NULL, blank=True, null=True, related_name="children"
+    )
```

Comment:
Wondering now if it'd be better for this to be PROTECT or CASCADE instead of SET_NULL...?